File: iOS-Drawing-Examples/03-DrawingImages/03-DrawingImages/ViewController.m
//
//  ViewController.m
//  03-DrawingImages
//
//  Created by ShannonChen on 2016/10/13.
//  Copyright © 2016年 YHouse. All rights reserved.
//

#import "ViewController.h"
#import "ImageViewController.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@property (strong, nonatomic) UITableView *tableView;
@property (strong, nonatomic) NSArray *dataSource;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.navigationItem.title = @"Drawing Images";
    
    self.view.backgroundColor = [UIColor whiteColor];
    self.dataSource = @[@"01-Creating a Custom Image",
                        @"02-Building a Thumbnail Image",
                        @"03-Extracting Portions of Images",
                        @"04-Converting an Image to Grayscale",
                        @"05-Watermarking Images",
                        @"06-Retrieving Image Data"];
    
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
    [self.view addSubview:self.tableView];
}


#pragma mark - <UITableViewDataSource>
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.dataSource.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    static NSString *identifier = @"identifier";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:identifier];
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:identifier];
    }
    cell.textLabel.text = self.dataSource[indexPath.row];
    
    return cell;
}

#pragma mark - <UITableViewDelegate>
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 44;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    ImageViewController *viewController = [[ImageViewController alloc] initWithIndex:indexPath.row];
    [self.navigationController pushViewController:viewController animated:YES];
}


@end

File: NSNotification/NSNotificationPlayground/NSNotificationPlayground/ViewController.swift
//
//  ViewController.swift
//  NSNotificationPlayground
//
//  Created by ShannonChen on 2017/9/5.
//  Copyright © 2017年 ShannonChen. All rights reserved.
//

import UIKit

class ViewController: UIViewController {
    
    static let notificationName = "notificationTest"
    

    override func viewDidLoad() {
        super.viewDidLoad()
        
        let red: CGFloat = CGFloat(arc4random() % 256) / 256.0
        let green: CGFloat = CGFloat(arc4random() % 256) / 256.0
        let blue: CGFloat = CGFloat(arc4random() % 256) / 256.0
        let alpha: CGFloat = CGFloat(arc4random() % 256) / 256.0
        view.backgroundColor = UIColor(red: red, green: green, blue: blue, alpha: alpha)
        
        NotificationCenter.default.addObserver(self, selector: #selector(didReceiveNotification), name: NSNotification.Name.init(ViewController.notificationName), object: nil)
    }

    
    /// 收到通知
    func didReceiveNotification() {
        print(title ?? "");
    }
}

